src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.Dynamic.cs

```diff
@@ -0,0 +1,289 @@
+        {
+            Debug.Assert(!_pendingTableSizeUpdate, "Dynamic table size update should be encoded before headers.");
+
+            // Never index sensitive value.
+            if (encodingHint == HeaderEncodingHint.NeverIndex)
+            {
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
+
+                return index == -1
+                    ? EncodeLiteralHeaderFieldNeverIndexingNewName(name, value, buffer, out bytesWritten)
+                    : EncodeLiteralHeaderFieldNeverIndexing(index, value, buffer, out bytesWritten);
+            }
+
+            // No dynamic table. Only use the static table.
+            if (_disableDynamicCompression || _maxHeaderTableSize == 0 || encodingHint == HeaderEncodingHint.IgnoreIndex)
+            {
+                return staticTableIndex == -1
+                    ? EncodeLiteralHeaderFieldWithoutIndexingNewName(name, value, buffer, out bytesWritten)
+                    : EncodeLiteralHeaderFieldWithoutIndexing(staticTableIndex, value, buffer, out bytesWritten);
+            }
+
+            // Header is greater than the maximum table size.
+            // Don't attempt to add dynamic header as all existing dynamic headers will be removed.
+            if (HeaderField.GetLength(name.Length, value.Length) > _maxHeaderTableSize)
+            {
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
+
+                return index == -1
+                    ? EncodeLiteralHeaderFieldWithoutIndexingNewName(name, value, buffer, out bytesWritten)
+                    : EncodeLiteralHeaderFieldWithoutIndexing(index, value, buffer, out bytesWritten);
+            }
+
+            return EncodeDynamicHeader(buffer, staticTableIndex, name, value, out bytesWritten);
+        }
+
+        private int ResolveDynamicTableIndex(int staticTableIndex, string name)
+        {
+            if (staticTableIndex != -1)
+            {
+                // Prefer static table index.
+                return staticTableIndex;
+            }
+
+            return CalculateDynamicTableIndex(name);
+        }
+
+        private bool EncodeDynamicHeader(Span<byte> buffer, int staticTableIndex, string name, string value, out int bytesWritten)
+        {
+            var headerField = GetEntry(name, value);
+            if (headerField != null)
+            {
+                // Already exists in dynamic table. Write index.
+                var index = CalculateDynamicTableIndex(headerField.Index);
+                return EncodeIndexedHeaderField(index, buffer, out bytesWritten);
+            }
+            else
+            {
+                // Doesn't exist in dynamic table. Add new entry to dynamic table.
+                var headerSize = (uint)HeaderField.GetLength(name.Length, value.Length);
+
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
+                var success = index == -1
```

Comment:
Minor nit: comparing against zero often leads to slightly tighter code.  Since these methods that return an index are all private, it's fine to do this:

```suggestion
                var success = index < 0
```